[Feat] 등록 장비의 데이터 수집을 위한 Batch 시스템 구축 (Device 상태 최신화)

IToriginal opened the pull request:
## #️⃣ Relationship Issues
- #16 

<br>

## 💡 Key Changes
- Spring Batch를 활용하여 지정 주기마다 데이터베이스의 상태를 확인합니다.
- Device 테이블에 등록된 DB의 정보를 통해 v$session 테이블을 조회할 수 있는 권한인지 판단하고 Device 테이블의 속성을 최신화합니다.
### Batch Step
1. Read: Device 장비를 모두 조회(JPA Repository 사용)
2. Processor: Device 장비 정보로 Data 접속 상태를 확인
3. Writer: Device status 상태 최신화

## Run Test
### 1️⃣ 등록된 장비와 테이블 컬럼의 상태는 같다.
- DataBase status: `OFF`
- Device Table status: `0(false)`
```
SessionDataScheduler     : >>> [ 📑 Run Collect Session Data ... ]
DeviceStatusReader       : >>> [ 🚀 Device Reader 초기화, 현재 등록 되어 있는 장비 수: 1 ]
DeviceStatusProcessor    : >>> [ 🚀 Device Processor 초기화 ]
DeviceStatusWriter       : >>> [ 🚀 Device Writer 초기화 ]

DeviceStatusReader       : >>> [ 📑 장비 정보 읽기: LOCALHOST ]
DeviceStatusReader       : >>> [ 🎯 더 이상 정보를 읽을 장비가 없습니다. ]

DeviceStatusProcessor    : >>> [ ❌ Device LOCALHOST 연결 실패. ... ]

DeviceStatusWriter       : >>> [ ⏭️ Device LOCALHOST 상태 변경 없음. 업데이트 생략 ]

DeviceStatusReader       : >>> [ ✅ 모든 장비 정보 읽기 완료 ]
DeviceStatusProcessor    : >>> [ ✅ 모든 장비 작업 완료 ]
DeviceStatusWriter       : >>> [ ✅ 모든 장비 쓰기 완료 ]
```

### 2️⃣ 등록된 장비와 테이블 컬럼의 상태가 다르다.
- DataBase status: `ON`
- Device Table status: `0(false)`
```
SessionDataScheduler     : >>> [ 📑 Run Collect Session Data ... ]
DeviceStatusReader       : >>> [ 🚀 Device Reader 초기화, 현재 등록 되어 있는 장비 수: 1 ]
DeviceStatusProcessor    : >>> [ 🚀 Device Processor 초기화 ]
DeviceStatusWriter       : >>> [ 🚀 Device Writer 초기화 ]

DeviceStatusReader       : >>> [ 📑 장비 정보 읽기: LOCALHOST ]
DeviceStatusReader       : >>> [ 🎯 더 이상 정보를 읽을 장비가 없습니다. ]

DeviceStatusProcessor    : >>> [ ✅ Device LOCALHOST 연결 성공 ]

DeviceStatusWriter       : >>> [ 🔄 Device LOCALHOST 상태 변경. 업데이트 상태: true ]

DeviceStatusReader       : >>> [ ✅ 모든 장비 정보 읽기 완료 ]
DeviceStatusProcessor    : >>> [ ✅ 모든 장비 작업 완료 ]
DeviceStatusWriter       : >>> [ ✅ 모든 장비 쓰기 완료 ]
```
> 업테이트 쿼리문이 실행 되어 상태 변경:  `0(false)` -> `1(true)`

<br>

## ➕ ETC 
- 현재는 Device 상태를 확인하고 상태를 업데이트 하는 Step이 생성이 되어 있습니다.
- 해당 Step은 이후에 데이터베이스의 상태가 true일 경우 데이터를 수집하도록 로직을 추가하여 개발할 예정입니다.
- if-else, method 분리 등 몇몇 가독성을 위한 코드는 이후에 다시 수정될 예정입니다. (로직이 추가되기 때문에 지금 변경해도 바뀜)
- Batch 시스템은 예외가 발생하게 되면 문제가 생겨서 Error 처리에 대해 고민이 되는 중입니다. (Exception 처리는 점차 수정하고 분리할 예정)
- 모듈 별 배너 변경: 실행 구분
![image](https://github.com/user-attachments/assets/dde74b67-48f7-4119-b634-539c371a65fe)
![image](https://github.com/user-attachments/assets/373f29eb-a675-478e-95f7-a9a58f0a1075)

IToriginal committed:
:sparkles: feat: SessionData Entity 생성

IToriginal committed:
:heavy_plus_sign: feat: Spring Batch Dependency 추가

IToriginal committed:
:hammer: fix: jpa 의존성 root 패키지로 이동

IToriginal committed:
:recycle: refactor: AES 암호화 클래스 경로 이동

IToriginal committed:
:recycle: refactor: AES256 암호화 클래스를 전역처리하지 않도록 수정

IToriginal committed:
:sparkles: feat: 컴포넌트, 빈 스캔 설정 및 batch meta-table 자동 생성 설정
- never: 생성하지 않음
- always: 항상 생성

IToriginal committed:
:sparkles: feat: Device 상태 Step Reader 클래스 생성

IToriginal committed:
:sparkles: feat: Device 상태 Step Processor 클래스 생성

IToriginal committed:
:sparkles: feat: JDBC Repository 및 Query 생성 클래스 생성

IToriginal committed:
:sparkles: feat: HikariDataSource 생성을 위한 유틸 클래스 구현

IToriginal committed:
:sparkles: feat: Device requestDto 생성

IToriginal committed:
:sparkles: feat: Device 상태 Step Writer 클래스 생성

IToriginal committed:
:sparkles: feat: Batch Job, Step(read, process, write) 설정 등록

IToriginal committed:
🔀 merge: #22, #20, #18 브랜치 병합

IToriginal committed:
:recycle: refactor: QueryLoader 패키기 경로 수정

IToriginal committed:
:recycle: refactor: Writer 메서드 분리

dabeen-jung commented on semo-core/src/main/java/sandbox/semo/domain/common/crypto/CryptoConfig.java:
이렇게 암호화와 관련된 것은 crypto를 이용해서 한 곳에서 관리하면 좋군요! 👍

IToriginal committed:
🔀 merge: #24 브랜치 병합

IToriginal committed:
:sparkles: feat: transaction 격리 설정

Caused by: oracle.jdbc.OracleDatabaseException: ORA-08177: 이 트랜잭션에 대한 직렬화 액세스를 할 수 없습니다

IToriginal committed:
:sparkles: feat: 모듈 별 로고 분리

dabeen-jung commented on semo-batch/src/main/java/sandbox/semo/batch/util/HikariDataSourceUtil.java:
case문을 통해서 타입별로 분리하셨는데 가독성이 좋네요!
배워갑니다 :)

dabeen-jung reviewed (approved):
LGTM 🎉

juneheel commented on semo-core/src/main/java/sandbox/semo/domain/collection/entity/SessionData.java:
```diff
@@ -0,0 +1,119 @@
+    @Column(name = "SQL_ID", length = 13)
+    private String sqlId;
+
+    @Column(name = "SQL_EXEC_START")
```
LocalDateTime이 아닌 Date를 사용하신 이유가 궁금합니다!

IToriginal replied:
해당 컬럼 docs type이 Date로 되어있어서 사용했습니다!

juneheel commented on semo-core/src/main/java/sandbox/semo/domain/collection/entity/SessionDataId.java:
```diff
@@ -0,0 +1,25 @@
+@Getter
+@Builder
+@AllArgsConstructor
+@EqualsAndHashCode
```
복합키에서는 `equals()와 hashCode()` 구현이 필수군요👍

juneheel reviewed (approved):
👍 어려운 부분이었는데 구현과 로깅을 너무잘해주셔서 이해하기 편했습니다. 수고많으셨어요! 특히 배너가 마음에듭니다 ㅎㅎ lgtm

taeyeon721 reviewed (approved):
LGTM🎃
